lib/tcp/pcb.ml

```diff
@@ -493,47 +457,49 @@ struct
-      | false, false ->
-        (* What the hell is this packet? No SYN,ACK,RST *)
-        Log.s debug "input-no-pcb: unknown packet";
-        Lwt.return_unit
+    match Tcp_parse.parse_tcp_header pkt with
+    | Result.Error s -> Log.debug (fun f -> f "parsing TCP header failed: %s" s);
+      Lwt.return_unit
+    | Result.Ok parsed ->
+      match parsed.rst with
+      | true -> process_reset t id ~ack:parsed.ack ~ack_number:parsed.ack_number
+      | false ->
+        match parsed.syn, parsed.ack with
+        | true , true  -> process_synack t id ~pkt ~ack_number:parsed.ack_number
+                            ~sequence:parsed.sequence ~tx_wnd:parsed.window
+                            ~options:parsed.options ~syn:parsed.syn ~fin:parsed.fin
+        | true , false -> process_syn t id ~listeners ~pkt ~tx_wnd:parsed.window
+                            ~ack_number:parsed.ack_number ~sequence:parsed.sequence
+                            ~options:parsed.options ~syn:parsed.syn ~fin:parsed.fin
+        | false, true  -> process_ack t id ~pkt ~ack_number:parsed.ack_number
+                            ~sequence:parsed.sequence ~syn:parsed.syn
+                            ~fin:parsed.fin
+        | false, false ->
+          (* No SYN, ACK, or RST, so just drop it *)
+          Log.debug (fun f -> f "incoming packet matches no connection table entry and has no useful flags set; dropping it");
+          Lwt.return_unit
 
   (* Main input function for TCP packets *)
   let input t ~listeners ~src ~dst data =
     match verify_checksum src dst data with
     | false ->
-      Log.s debug "RX.input: checksum error";
+      Log.debug (fun f -> f "TCP checksum was not correct; discarding packet");
       Lwt.return_unit
     | true ->
-      let source_port = Tcp_wire.get_tcp_src_port data in
-      let dest_port = Tcp_wire.get_tcp_dst_port data in
-      let id =
-        { WIRE.local_port = dest_port;
-          dest_ip         = src;
-          local_ip        = dst;
-          dest_port       = source_port }
-      in
-      (* Lookup connection from the active PCB hash *)
-      with_hashtbl t.channels id
-        (* PCB exists, so continue the connection state machine in tcp_input *)
-        (Rx.input t data)
-        (* No existing PCB, so check if it is a SYN for a listening function *)
-        (input_no_pcb t listeners data)
+      try
+        let source_port = Tcp_wire.get_tcp_src_port data in
+        let dest_port = Tcp_wire.get_tcp_dst_port data in
+        let id = WIRE.wire ~local_port:dest_port ~dest_port:source_port ~dest_ip:src ~local_ip:dst in
+        (* Lookup connection from the active PCB hash *)
+        with_hashtbl t.channels id
+          (* PCB exists, so continue the connection state machine in tcp_input *)
+          (Rx.input t data)
+          (* No existing PCB, so check if it is a SYN for a listening function *)
+          (input_no_pcb t listeners data)
+      with
+      | Invalid_argument s ->
```

Comment:
On looking at this again it's not clear to me why this is in its own `try` at all; I've added the ports to the record returned by the parse function and now `input` just runs `parse` right away.  Thanks for noticing this.